### DISCUSSION TITLE
`useSubscription` doesn't get React state updates

### DISCUSSION THREAD

**Discussion by gustavo-shigueo** [trpc/trpc#4731] [Q&A]
I have a ContextProvider with some state which uses a web socket subscription through the `useSubscription` method. The isue is, the `onData` method depends on the Provider's state, but it doesn't get updated, only ever getting the initial state passed into the `useState` function

```jsx
function Component() {
    const [state, setState] = useState('initial')
    const triggerSocket = trpc.trigger.useMutation()

    trpc.socket.useSubscription({
        onData() {
            console.log(state) // initial
        }
    })

    return <button onClick={async () => {
        setState('updated')
        await triggerSocket.mutateAsync()
    }}>Click me</button>
}
```

**Comment by gustavo-shigueo** [comment]
I managed to work around the issue with `useRef`. Still, if someone figures out a fix for my `useState` version, feel free to leave it here, as it may help someone else in the future

```jsx
function Component() {
    const ref = useRef('initial')
    const triggerSocket = trpc.trigger.useMutation()

    trpc.socket.useSubscription({
        onData() {
            console.log(ref.current) // updated
        }
    })

    return <button onClick={async () => {
        ref.current = 'updated'
        await triggerSocket.mutateAsync()
    }}>Click me</button>
}
```

**Comment by escritorio-gustavo** [comment]
I also have this issue, and I really  need to use state rather than ref due to needing a component to rerender on every change. As a workaround (a.k.a a really disgusting hack) I'm using a state and a ref at the same time, making sure to keep them in sync... Which is horrible because it's sooo easy to screw up (I had two separate bugs due to messing this up)

**Comment by infodusha** [comment]
That seems to be fixed now in https://github.com/trpc/trpc/commit/2e2df9ef5a9f6a34c9380c0a3504b490cfb7ba6b
